OptionParser rewrite with subcommands

RX14 opened the pull request:
I got angry with `OptionParser` for being a pile of bad code so I rewrote the implementation. It's simpler now, and easier to understand the implementation, and not `O(n^2)` so probably miles faster. And it has a completely compatible interface, no changes were required to any existing usages across the stdlib.

While I was rewriting it I had a universe brain moment and realised that I could make it handle subcommands very simply, so I threw that in as a bonus.

Example subcommand usage:
```cr
OptionParser.parse(args) do |opts|
  opts.on("subcommand", "Description") do
    opts.on("--foo arg", "Foo") { |v| foo = v }
    opts.on("--bar", "Bar") { bar = true }
    opts.on("-z", "--baz", "Baz") { z = true }
  end
  opts.on("--verbose", "") { verbose = true }
end
```

There's some tweaks still required to get the `to_s` output to not suck when using subcommands but I wanted to submit this for comment before I forgot about it tomorrow morning.

Fixes #8937.

RX14 commented on spec/std/option_parser_spec.cr:
```diff
@@ -101,8 +101,8 @@ describe "OptionParser" do
-  it "has required option with = (3) raises" do
-    expect_missing_option ["--flag="], "--flag=FLAG", "--flag"
+  it "has required option with = (3) handles empty" do
+    expect_capture_option ["--flag="], "--flag=FLAG", ""
```
I changed the behaviour of `OptionParser` here because I thought this behaviour was stupid.

oprypin commented:
I think this fixes https://github.com/crystal-lang/crystal/issues/8937

RX14 commented:
Yeah, me too

I'll add a spec.

straight-shoota commented:
Awesome!

This might also enable to implement #5338 (could be a follow-up though).

RX14 commented:
Ah, that seems rather difficult with the current code because there's no way to "peek" at the next argument and see if it's valid or not

I might just get  rid of the "iterator" business and use indexing in the array, then.

straight-shoota commented:
I suppose it might be best to support only `--optional-flag=VALUE` or `--optional-flag` (no VALUE), an not `--optional-flag VALUE` because the latter is always kind of ambiguous, even with a fixed set of values and it can't handle arbitrary values at all.

RX14 commented:
It's definitely doable to have a "next argument doesn't start with -" heuristic.

I switched the code to use array indexing and it's neater anyway.

Blacksmoke16 commented:
Thoughts on allowing multiple shortflags to be grouped?

```cr
require "option_parser"

OptionParser.parse do |opts|
  opts.on("-f", "Prints foo") { pp "foo" }
  opts.on("-b", "Prints bar") { pp "bar" }
end
```

Works:
`./test -f -b`

Does not work:
`./test -fb`

RX14 commented:
@Blacksmoke16 I thought about adding that feature, but I'd rather leave it for another PR.

Blacksmoke16 commented:
Would also close #3357.

Another thought I had, probably better for another PR; but what about parsing options with more than one value?  An example of this is the `--arg name value` within https://stedolan.github.io/jq/manual/.

oprypin commented:
@Blacksmoke16  It is indeed unrelated and is also not part of any spec or common use.

oprypin commented:
I wrote this reply before realizing that something else was being asked, but

if one wants to support an arbitrary number of values, it should definitely not be done as
`--stuff foo bar baz` (non-compliant and very ambiguous in terms of what can follow it)
but rather `--stuff foo --stuff=bar --stuff=baz`,
or, as a shell would let you write, `--stuff={foo,bar,baz}`.

And this is already supported, just push it to an array:
<details>
<summary>Code</summary>

```crystal
require "option_parser"

argv = ["--stuff", "foo", "--stuff=bar"]

stuff = [] of String

OptionParser.parse(argv) do |parser|
  parser.on("--stuff=NAME", "") { |name| stuff << name }
end

p! stuff # => ["foo", "bar"]
```

</details>

Blacksmoke16 commented:
@oprypin I wasn't imagining supporting an arbitrary number of values, just a way to handle a fixed number that is more than one.

Like `parser.on("--arg NAME VALUE", "Add an arg") { |name, value| # Do whatever }`

oprypin commented:
>>>>I wrote this reply before realizing that something else was being asked, but

straight-shoota commented:
@Blacksmoke16 IMO that looks like a very uncommon feature. I'd rather not have such specialized things in OptionParser.

Btw. why not just use `--arg NAME=VALUE` for example? Would be much more convenient IMO.

Blacksmoke16 commented:
Oh I don't really have a use case.  Was just something I noticed in `jq`.

sudo-nice commented:
Thank you for the great improvement of the OptionParser!
Will those sticky arguments (like `command -ufoo -tbar`) still be working after the update?
<details>
  <summary>Currently working example</summary>

  ```crystal
  require "option_parser"

  ary = [] of String
  options = %w[-ufoo -tbar]

  OptionParser.parse(options) do |p|
    p.on("-u FOO", "Description") { |o| ary << o }
    p.on("-t FOO", "Description") { |o| ary << o }
  end

  pp ary # => ["foo", "bar"]
  ```
</details>

oprypin commented:
You can just copy paste [the file](https://github.com/RX14/crystal/blob/feature/option-parser-rewrite/src/option_parser.cr) before your example to try.

Yes, it works before and after.

RX14 commented:
I've just added commits to unregister subcommands when they're called, this means that when you call a subcommand `foo`, *all* subcommands are unregistered, so `foo foo` doesn't handle `foo` twice.

I've also added two new features: `OptionParser#stop` and `OptionParser#every(&)`. The latter takes a callback to run on every single argument before it's parsed. This is neccesary, because performing a `map` or `each` on `ARGV` wouldn't be able to tell the difference between flags and their arguments, whereas `every` is not passed flag arguments. Then `OptionParser#stop` allows you to bail out of parsing at any point.

straight-shoota commented on spec/std/option_parser_spec.cr:
```diff
@@ -438,4 +442,127 @@ describe "OptionParser" do
+    OptionParser.parse(args) do |opts|
+      opts.on("--foo", "") { foo = true }
+      opts.on("--bar", "") { bar = true }
+      opts.every do |arg|
```
I find it a bit confusing that calling `opts.stop` in `every` block behaves differently than in `on` block (/re whether the current arg is consumed). Maybe renaming it to `before_every` would help. Or `before_each`?

RX14 replied:
yeah, before_each sounds good for me.

RX14 commented on src/option_parser.cr:
```diff
@@ -201,141 +228,134 @@ class OptionParser
+        # subcommands since they are no longer valid.
+        unless flag.starts_with?('-')
+          @handlers.select! { |k, v| k.starts_with?('-') }
+          @flags.select! { |flag| flag.starts_with?("    -") }
```
Check out https://github.com/crystal-lang/crystal/blob/7e2e84042a393223660a216fbc5d0d5cd6bb80a5/src/option_parser.cr#L178

RX14 commented:
`every` has been renamed to `before_each` and documentation is complete. This is ready for final review.

straight-shoota commented:
I'm not entirely sure about using the same method for defining a subcommands and flags when the only difference is whether the string starts with a dash.
Maybe a different method name for sucommand would be better to more clearly communicate the different concepts. For example `opts.subcommand("subcommand", "") do`.

RX14 commented:
I mean, we could have different methods for short and long flags because they have differet behaviours too.

I think this is fine.

didactic-drunk commented:
Is reuse of an `OptionParser` allowed with new data?  I have a repl loop that reuses `OptionParser` objects for it's command interface.

Does this handle nested subcommands? I have a program that has approximately 80 commands total up to 4? subcommands deep.  Each command has it's options parsed at the subcommand level.
```
-v foo --log=log.txt bar --drinks=2 baz 
-v foo --log=log.txt --drinks=2 bar baz  # Invalid.  --drinks is bar scope only.
-v foo bar --log=log.txt --drinks=2 baz  # Invalid.  --log is foo scope only.
foo -v --log=log.txt bar --drinks=2 baz  # Invalid.  -v is global scope only.
```

Will this accomodate?

RX14 commented:
@didactic-drunk you can't reuse an optionparser after it's parsed subcommands because it's mutated the handlers list.

Nested subcommands work fine. `--flags` are supported *after* the subcommand which registers them. So your last two examples would be valid, and this is fine.

This is not intended to do everything. This is intended to solve common problems. Whatever you're doing doesn't seem to be typical...

RX14 commented:
Please can I have another review, I have a PR waiting on this...

asterite commented:
> @didactic-drunk you can't reuse an optionparser after it's parsed subcommands because it's mutated the handlers list

I think this should be modified. The current optionparser (not the one in this PR) creates a `ParseTask` and runs that so that it doesn't modify the original option parser.

@RX14 I think doing the same would be easy: just clone all the handlers and everything and pass it to another class that does the parsing. That way it'll also be easier to understand which class is in charge of the configuration and which is in charge of the parsing.

straight-shoota commented:
> You should be able to reuse an OptionParser

Why? What's the use case for that? OptionParser is mostly used in CLI applications that just need to parse the program arguments once.

If you want to implement a REPL or similar that needs repeated parsing, you can just configure the option parser and clone it before doing the actual parse. It should be fine to require this simple extra step for such advanced use cases. It keeps the overall complexity low.

asterite commented:
> Why?

Because in my mind it doesn't make sense that if I use OptionParser twice I get two different results. Plus it's a breaking change. And I'm sure it works like that in Ruby too.

RX14 commented:
I don't think this is possible, because of the way `OptionParser` implements subcommands:

```cr
OptionParser.parse(args) do |opts|
  opts.on("subcommand", "") do
    opts.on("--foo arg", "") { |v| foo = v }
    opts.on("--bar", "") { bar = true }
    opts.on("sub2", "") { sub2 = true }
  end
  opts.on("--verbose", "") { verbose = true }
  opts.on("-z", "--baz", "") { z = true }
end
```

The `OptionParser` is mutated while parsing using `opts.on` to register new flags.

It'd take a redesign of the interface to support this, and I'm personally not sure it's worth it. It's not a breaking change either, because the only time `OptionParser` is mutated is when doing subcommands, which we didn't support before anyway.

asterite commented:
I think this can be implemented without mutation. Until that's done, sorry, I'm not approving this.

didactic-drunk commented:
Couldn't mutation be solved by passing a new `OptionParser` object to the subcommand block?

```
OptionParser.parse(args) do |opts|
  opts.on("subcommand", "") |subopts| do
  end
end
```

straight-shoota commented:
@asterite Do you think there's anything wrong with explicitly using `clone` if you want to reuse an instance?

asterite commented:
> Do you think there's anything wrong with explicitly using clone if you want to reuse an instance?

Yes, it's unexpected. OptionParser defines how options should be parsed. I find it pretty strange that if I use it twice, it behaves differently.

asterite commented:
Crystal is not a functional programming language, but we tend to use and like purity. That is, methods shouldn't have unexpected side effects.

RX14 commented:
I made `#parse` restore state on exit.

I tried going with the route of cloning the `OptionParser` and creating a DSL like @didactic-drunk suggested, but it's not neat to synchronise `@stop`, and the `@before_each` handlers etc. between the *original* instance which is passed to the `OptionParser.new do |parser|` block, and the instance which is passed to `parser.subcommand("foo", "") do |parser|`. Basically you need to unify all the instance variables between the two *except* `@flags` and `@handlers`, which is much uglier than what I just comitted.

didactic-drunk commented:
Can you have an additional `OptionParser` per subcommand and start parsing ARGV from the leftovers?  Doesn't the parent OptionParser stop when a subcommand is reached (eliminating the @stop problelm)?  Perhaps @asterite's state suggestion would solve the issue?

`def parse(args = ARGV, state = State.new)`

asterite commented:
I guess with that proof we can go ahead without the last commit, and add it later if needed.

RX14 commented:
@didactic-drunk in short, no.

RX14 commented:
We might as well leave the last commit. Why not?

Sija commented on src/option_parser.cr:
```diff
@@ -417,6 +425,15 @@ class OptionParser
         @invalid_option.call(arg)
       end
     end
+  ensure
+    @flags = old_flags.not_nil!
+    @handlers = old_handlers.not_nil!
+    @stop = false
+    @banner = old_banner
+    @unknown_args = old_unknown_args
+    @missing_option = old_missiong_option.not_nil!
+    @invalid_option = old_invalid_option.not_nil!
+    @before_each = old_before_each
```
Wouldn't be better to use `begin ... end` block instead of using `.not_nil!`?

RX14 replied:
I don't think it's worth the extra indent. They're semantically equivalent anyway.

oprypin replied:
OK then how about clumping it all into one variable (tuple? record? maybe store the whole state in a sub-record in the first place?)

straight-shoota replied:
It should be either as is or without the reset.

Sija replied:
@RX14 Using `.not_nil!` should be treated as a last resort, so if there's a way to avoid it, even at the "cost" of an extra indent, then I'd say it's worth it.

asterite replied:
Just put the begin after you save all those variables. I agree with Sija.

RX14 commented:
I am really ambivalent about leaving the last commit in or not, since it's less work to do so, I'll leave it in.

I'll merge this today if there's no more API-level objections.

Sija commented on src/option_parser.cr:
```diff
@@ -201,141 +293,151 @@ class OptionParser
+    old_handlers = @handlers.clone
+    old_banner = @banner
+    old_unknown_args = @unknown_args
+    old_missiong_option = @missing_option
```
```suggestion
    old_missing_option = @missing_option
```

waj commented:
@RX14 thanks for working on this. I agree that the current `OptionParser` is bad designed and buggy. This implementation is better than the one we currently have and from that perspective I have no objections to merge this PR.

Now, I wouldn't consider this implementation "good code". I don't need to check usage stats to understand that the approach used for subcommands is not well designed. Mutating the parser is unnecessary and TBH relying on string comparison with a specific number of spaces from a formatted text to do so, makes me cringe.

I think it could actually just make it store the options more structured, format those options just when it's needed (on `--help`) and filter based on those structures, not strings.

Also, I wish the subcommands implementation could work somehow with separate `OptionParser` instances. That way subcommands could be defined in separate files, or make them join/split to different processes more easily. For example, take a look at how it's currently used for Crystal compiler subcommands, and think how this new implementation would fit there.

I could understand that you don't feel like doing it right now, but I wouldn't say this PR is the ultimate design and good code that we could be looking for.

didactic-drunk commented:
> Also, I wish the subcommands implementation could work somehow with separate `OptionParser` instances. That way subcommands could be defined in separate files, or make them join/split to different processes more easily.

Should subcommands be marked experimental?  Should `opts.on` be `opts.subcommand` or some other method to make the meaning clear?

RX14 commented:
> I could understand that you don't feel like doing it right now, but I wouldn't say this PR is the ultimate design and good code that we could be looking for.

On that, I totally agree. The `--help` text generation needs to be refactored. The rest isn't so bad. But this is still better than before.
